x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,404 @@
+        if (false == Files.getOwner(instantAutoConfigDir, LinkOption.NOFOLLOW_LINKS).equals(Files.getOwner(env.configFile(),
+                LinkOption.NOFOLLOW_LINKS))) {
+            Files.deleteIfExists(instantAutoConfigDir);
+            throw new UserException(ExitCodes.CONFIG, "Aborting auto configuration because it would change file owners");
+        }
+
+        // the transport key-pair is the same across the cluster and is trusted without hostname verification (it is self-signed),
+        final X500Principal certificatePrincipal = new X500Principal(buildDnFromDomain(System.getenv("HOSTNAME")));
+        final GeneralNames subjectAltNames = getSubjectAltNames();
+
+        KeyPair transportKeyPair = CertGenUtils.generateKeyPair(TRANSPORT_KEY_SIZE);
+        // self-signed which is not a CA
+        X509Certificate transportCert = CertGenUtils.generateSignedCertificate(certificatePrincipal,
+                subjectAltNames, transportKeyPair, null, null, false, TRANSPORT_CERTIFICATE_DAYS, null);
+        KeyPair httpCAKeyPair = CertGenUtils.generateKeyPair(HTTP_CA_KEY_SIZE);
+        // self-signed CA
+        X509Certificate httpCACert = CertGenUtils.generateSignedCertificate(certificatePrincipal,
+                subjectAltNames, httpCAKeyPair, null, null, true, HTTP_CA_CERTIFICATE_DAYS, null);
+        KeyPair httpKeyPair = CertGenUtils.generateKeyPair(HTTP_KEY_SIZE);
+        // non-CA
+        X509Certificate httpCert = CertGenUtils.generateSignedCertificate(certificatePrincipal,
+                subjectAltNames, httpKeyPair, httpCACert, httpCAKeyPair.getPrivate(), false, HTTP_CERTIFICATE_DAYS, null);
+
+        Path httpCAPath = instantAutoConfigDir.resolve(HTTP_AUTOGENERATED_CA_NAME + ".pem");
+        fullyWriteFile(httpCAPath, stream -> {
+            try (JcaPEMWriter pemWriter = new JcaPEMWriter(new BufferedWriter(new OutputStreamWriter(stream, StandardCharsets.UTF_8)))) {
+                pemWriter.writeObject(httpCACert);
+            }
+        });
+
+        try (SecureString nodeKeystorePassword = new SecureString(terminal.readSecret("", KeyStoreWrapper.MAX_PASSPHRASE_LENGTH));
+             KeyStoreWrapper nodeKeystore = KeyStoreWrapper.bootstrap(env.configFile(), () -> nodeKeystorePassword)) {
+            // do not overwrite keystore entries, instead expect the user to manually remove them himself if they've been generated by a
+            // previous invocation of this tool
+            if (nodeKeystore.getSettingNames().contains("xpack.security.transport.ssl.keystore.secure_password") ||
+                nodeKeystore.getSettingNames().contains("xpack.security.transport.ssl.truststore.secure_password") ||
+                nodeKeystore.getSettingNames().contains("xpack.security.http.ssl.keystore.secure_password")) {
+                throw new UserException(ExitCodes.CONFIG, "Aborting auto configuration because the node keystore contains password " +
+                        "settings already");
+            }
```

Comment:
Would it make sense to check these settings and bail out before we generate the certs?

Reply:
I think it might make sense to keep the generation part separate from the writing, but I'd be fine eitherway. This is however part of #74868 and not this PR, so please add this comment in your review there